pallets/proxy-keystore/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+        /// Any entry that matches the `key_hash` and the key purpose in the `LastKeyByPurpose`
+        /// storage is removed.
+        ///
+        /// The key deposit is returned upon success.
+        fn revoke_key(
+            account_id: T::AccountId,
+            key_hash: T::Hash,
+            block_number: T::BlockNumber,
+        ) -> DispatchResult {
+            let mut key_id_opt: Option<KeyId<T::Hash>> = None;
+
+            let iter = <Keys<T>>::iter_prefix(account_id.clone());
+
+            // Search for a key that has the key_hash that we are looking for.
+            for (key_id, _) in iter {
+                if key_id.0 == key_hash {
+                    key_id_opt = Some(key_id);
+                    break
+                }
+            }
+
+            return match key_id_opt {
+                Some(key_id) => {
+                    // Retrieve the key from storage.
+                    // TODO(cdamian): Given the above, are there any chances of a race here?
+                    <Keys<T>>::try_mutate(
+                        account_id.clone(),
+                        key_id,
+                        |key_opt| -> DispatchResult  {
+                            return match key_opt {
+                                Some(key) => {
+                                    // Check if key was already revoked.
+                                    if let Some(_) = key.revoked_at {
+                                        return Err(Error::<T>::KeyRevoked.into())
+                                    }
+
+                                    // Revoke it at the current block number.
+                                    key.revoked_at = Some(block_number);
+
+                                    // Check if we have a key by purpose that matches our key_hash.
+                                    <LastKeyByPurpose<T>>::try_mutate(
+                                        account_id.clone(),
+                                        key.purpose.clone(),
+                                        |last_key_id_opt| -> DispatchResult {
+                                            return match last_key_id_opt {
+                                                Some(last_key_id) => {
+                                                    if last_key_id.0 == key_hash {
+                                                        // Key by purpose found, clear it.
+                                                        *last_key_id_opt = None;
+                                                    }
+
+                                                    Ok(())
+                                                },
+                                                None => Ok(()),
+                                            }
+                                        },
+                                    )?;
+
+                                    // Return the deposit.
+                                    Self::return_key_deposit(account_id, key.deposit);
```

Comment:
Since the key still is in storage, only in revoked state, we cannot return yet the deposit.

Reply:
OK, I removed this.